pkg/action/sysext.go

```diff
@@ -0,0 +1,356 @@
+
+const (
+	sysextDir        = "/var/lib/kairos/extensions/"
+	sysextDirActive  = "/var/lib/kairos/extensions/active"
+	sysextDirPassive = "/var/lib/kairos/extensions/passive"
+	sysextEnabledDir = "/run/extensions/"
+)
+
+// SysExtension represents a system extension
+type SysExtension struct {
+	Name     string
+	Location string
+}
+
+func (s *SysExtension) String() string {
+	return s.Name
+}
+
+// ListSystemExtensions lists the system extensions in the given directory
+// If none is passed then it shows the generic ones
+func ListSystemExtensions(cfg *config.Config, bootState string) ([]SysExtension, error) {
+	switch bootState {
+	case "active":
+		cfg.Logger.Debug("Listing active system extensions")
+		return getDirExtensions(cfg, sysextDirActive)
+	case "passive":
+		cfg.Logger.Debug("Listing passive system extensions")
+		return getDirExtensions(cfg, sysextDirPassive)
+	default:
+		cfg.Logger.Debug("Listing all system extensions (Enabled or not)")
+		return getDirExtensions(cfg, sysextDir)
+	}
+}
+
+// getDirExtensions lists the system extensions in the given directory
+func getDirExtensions(cfg *config.Config, dir string) ([]SysExtension, error) {
+	var out []SysExtension
+	// get all the extensions in the sysextDir
+	entries, err := cfg.Fs.ReadDir(dir)
+	// We don't care if the dir does not exist, we just return an empty list
+	if err != nil && !os.IsNotExist(err) {
+		return nil, err
+	}
+	for _, entry := range entries {
+		if !entry.IsDir() && filepath.Ext(entry.Name()) == ".raw" {
+			out = append(out, SysExtension{Name: entry.Name(), Location: filepath.Join(dir, entry.Name())})
+		}
+	}
+	return out, nil
+}
+
+// GetSystemExtension returns the system extension for a given name
+func GetSystemExtension(cfg *config.Config, name, bootState string) (SysExtension, error) {
+	// Get a list of all installed system extensions
+	installed, err := ListSystemExtensions(cfg, bootState)
+	if err != nil {
+		return SysExtension{}, err
+	}
+	// Check if the extension is installed
+	// regex against the name
+	re, _ := regexp.Compile(name)
+	for _, ext := range installed {
+		if re.MatchString(ext.Name) {
+			return ext, nil
+		}
+	}
+	// If not, return an error
+	return SysExtension{}, fmt.Errorf("system extension %s not found", name)
+}
+
+// EnableSystemExtension enables a system extension that is already in the system for a given bootstate
+// It creates a symlink to the extension in the target dir according to the bootstate given
+// It will create the target dir if it does not exist
+// It will check if the extension is already enabled but not fail if it is
+// It will check if the extension is installed
+func EnableSystemExtension(cfg *config.Config, ext string, bootState string) error {
+	// first check if the extension is installed
+	extension, err := GetSystemExtension(cfg, ext, "")
+	if err != nil {
+		return err
+	}
+
+	var targetDir string
+	switch bootState {
+	case "active":
+		targetDir = sysextDirActive
```

Comment:
mmm shouldn't the symlink go to `/run/sysextensions`?  what am I missing here?

Reply:
you are missing the whole thing :D

This only creates symlinks into /var/lib/kairos/extensions/active/passive to mark those extensions as "enabled" for a boot state.

Then its immucore task on bootup to check those and copy them into /run/extensions during initramfs (or even link them again? Can you do that?) so when systemd-sysext service is up it founds those.

This means that we dont have to deal with /var/lib/extensions in persistent between boots (if you enable X in active, it will appear also on passive as they share persistent) and can manage them separately in our own dirs.

Reply:
There, now you can pass a `--now` flag to the enable command and if the extension you are enabling for a bootstate matches the current bootstate (so you are enabling X extension for active and you are on active) it will symlink it to /run/extensions and refresh the extensions, so it can be used inmediatly

Reply:
ok cool - wasn't entirely clear neither from the PR/changeset that another piece of this its required. Any reason to use symlinks at this point? wouldn't be more clean to have a statefile that immucore reads?

Reply:
I thought so at the start but it would end up basically as a database in a file. Symlinks are universal, you don't need specialized tools to create them or anything else, just a shell and it requires no deps to manage them. Plus systemd docs mention that it possible you should use symlinks to add syexts to the sysext directories so it felt natural to use those directly. Not much stuff that can go wrong with them, they are either there and valid or invalid and that's basically it, allows to maybe grow easily for specific boot states that we may want in the future and so on 

Having a statefile presented more issues that that in the future I think. Maybe not, I just wanted to make it as simple and accesible as possible

Reply:
This way, immucore just read the dir a cording to the current boot then just goes over each entry enabling stuff.

That's it. No more to it.

Reply:
Also think about this, you could have a different disk where you store your extensions and manually symlink them. done, simple, no need for other tools. Could even do so on cloud-init stuff before sysext is run. No need for the code that checks that statefile to be in both places, just simple file managing, solved problem for 3000 years :P